internal/telemetry/option.go

```diff
@@ -0,0 +1,153 @@
+	for _, opt := range opts {
+		opt(c)
+	}
+}
+
+// WithNamespace sets name as the telemetry client's namespace (tracer, profiler, appsec)
+func WithNamespace(name Namespace) Option {
+	return func(client *Client) {
+		client.Namespace = name
+	}
+}
+
+// WithEnv sets the app specific environment for the telemetry client
+func WithEnv(env string) Option {
+	return func(client *Client) {
+		client.Env = env
+	}
+}
+
+// WithService sets the app specific service for the telemetry client
+func WithService(service string) Option {
+	return func(client *Client) {
+		client.Service = service
+	}
+}
+
+// WithVersion sets the app specific version for the telemetry client
+func WithVersion(version string) Option {
+	return func(client *Client) {
+		client.Version = version
+	}
+}
+
+// WithHTTPClient specifies the http client for the telemetry client
+func WithHTTPClient(httpClient *http.Client) Option {
+	return func(client *Client) {
+		client.Client = httpClient
+	}
+}
+
+func defaultAPIKey() string {
+	return os.Getenv("DD_API_KEY")
+}
+
+// WithAPIKey sets the DD API KEY for the telemetry client
+func WithAPIKey(v string) Option {
+	return func(client *Client) {
+		client.APIKey = v
+	}
+}
+
+// WithURL sets the URL for where telemetry information is flushed to.
+// For the URL, uploading through agent goes through
+//
+//	${AGENT_URL}/telemetry/proxy/api/v2/apmtelemetry
+//
+// for agentless:
+//
+//	https://instrumentation-telemetry-intake.datadoghq.com/api/v2/apmtelemetry
+//
+// with an API key
+func WithURL(agentless bool, agentURL string) Option {
+	return func(client *Client) {
+		if agentless {
+			client.URL = getAgentlessURL()
+		} else {
+			u, err := url.Parse(agentURL)
+			if err == nil {
+				u.Path = "/telemetry/proxy/api/v2/apmtelemetry"
+				client.URL = u.String()
+			} else {
+				client.log("Agent URL %s is invalid, switching to agentless telemetry endpoint", agentURL)
+				client.URL = getAgentlessURL()
+			}
+		}
+	}
+}
+
+func getAgentlessURL() string {
+	agentlessEndpointLock.RLock()
+	defer agentlessEndpointLock.RUnlock()
+	return agentlessURL
+}
+
+// configEnvFallback returns the value of environment variable with the
+// given key if def == ""
+func configEnvFallback(key, def string) string {
+	if def != "" {
+		return def
+	}
+	return os.Getenv(key)
+}
+
+// applyFallbackOps applies default values to the client unless
+// those values are already set.
+func (c *Client) applyFallbackOps() {
+	if c.Client == nil {
+		WithHTTPClient(defaultHTTPClient)(c)
+	}
+	if len(c.APIKey) == 0 && c.URL == getAgentlessURL() {
+		WithAPIKey(defaultAPIKey())(c)
+		if c.APIKey == "" {
+			c.log("Agentless is turned on, but valid DD API key was not found.")
+		}
+	}
+	c.Service = configEnvFallback("DD_SERVICE", c.Service)
+	if len(c.Service) == 0 {
+		if name := globalconfig.ServiceName(); len(name) != 0 {
+			c.Service = name
+		} else {
+			// I think service *has* to be something?
+			c.Service = "unnamed-go-service"
```

Comment:
let's check with the instrumentation telemetry folks on this

Reply:
I was basing this on the documentation for the `application` object in the V1 spec, which marks the `service_name` field as required. I forgot at this point if I confirmed this in practice. But I *think* based on digging through the telemetry backend code that it doesn't actually need to be set. I'm okay with it being up to the tracer library to set this.

Reply:
I noticed the profiler and tracer both have `filepath.Base(os.Args[0])` as the default service name in case the service name is not available from any other source, so i also just added that here